### ISSUE TITLE
ComputationGraph (CNN->RNN) runs on BackpropType.Standard but fails with TruncatedBPTT

### ISSUE THREAD

**Issue by Mioh** [deeplearning4j/deeplearning4j#2899]
While trying out a scaled down language model inspired by [this paper](https://arxiv.org/pdf/1508.06615.pdf) I noticed that` BackpropType.Standard` runs fine while `BackpropType.TruncatedBPTT` raises and Exception on D4J 0.7.2.

**Log**
```
15:17:52.378 [main] WARN org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer - Early stopping training terminated due to exception at epoch 0, iteration 0
org.deeplearning4j.exception.DL4JInvalidInputException: Received input with size(1) = 64 (input array shape = [200, 64, 2, 1]); input.size(1) must match layer nIn size (nIn = 128)
	at org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.activateHelper(LSTMHelpers.java:142)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.activateHelper(GravesLSTM.java:150)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.rnnActivateUsingStoredState(GravesLSTM.java:222)
	at org.deeplearning4j.nn.graph.ComputationGraph.rnnActivateUsingStoredState(ComputationGraph.java:2081)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:955)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:151)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:54)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:51)
	at org.deeplearning4j.nn.graph.ComputationGraph.doTruncatedBPTT(ComputationGraph.java:2032)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:824)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:798)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:660)
	at org.deeplearning4j.earlystopping.trainer.EarlyStoppingGraphTrainer.fit(EarlyStoppingGraphTrainer.java:74)
	at org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer.fit(BaseEarlyStoppingTrainer.java:109)
	at NeuralLanguageModelTrainer.Trainer.train(Trainer.java:101)
	at NeuralLanguageModelTrainer.Trainer.main(Trainer.java:74)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
```

**Architecture**
```
public ComputationGraphConfiguration smallCnnLstm() {
// CNN image dimensions
int numEncodings = 1;       // depth
int sequenceLength = 6;    // height
int maxWordLength = 20;    // width
// LSTM Properties
int lstmLayerSize = 250;
int tbpttLength = 35;
return new NeuralNetConfiguration.Builder()
                .optimizationAlgo(OptimizationAlgorithm.STOCHASTIC_GRADIENT_DESCENT)
                .seed(1234)
                .l2(0.001)
                .regularization(true)
                .rmsDecay(0.95)
                .learningRate(learningRate)
                .graphBuilder()
                .addInputs("input")
                .setInputTypes(InputType.convolutional(sequenceLength,maxWordLength,numEncodings))
                .addLayer("cnn1", new ConvolutionLayer.Builder(2, maxWordLength)
                        .stride(3, maxWordLength)
                        .nIn(1)
                        .nOut(32)
                        .activation(Activation.TANH)
                        .build(), "input")
                .addLayer("cnn2", new ConvolutionLayer.Builder(3, maxWordLength)
                        .stride(2, maxWordLength)
                        .nIn(1)
                        .nOut(32)
                        .activation(Activation.TANH)
                        .build(), "input")
                .addLayer("max1", new SubsamplingLayer.Builder(SubsamplingLayer.PoolingType.MAX)
                        .kernelSize(1, 1)
                        .stride(1,1)
                        .build(), "cnn1", "cnn2")
                .addLayer("lstm1", new GravesLSTM.Builder()
                        .nOut(lstmLayerSize)
                        .updater(Updater.RMSPROP)
                        .activation(Activation.RELU)
                        .weightInit(WeightInit.RELU)
                        .build(),"max1")
                .addLayer("output", new RnnOutputLayer.Builder(LossFunctions.LossFunction.MCXENT)
                        .activation(Activation.SOFTMAX)
                        .updater(Updater.RMSPROP)
                        .nIn(lstmLayerSize)
                        .nOut(iterator.totalOutcomes())
                        .build(),"lstm1")
                .setOutputs("output")
                .backprop(true)
                .backpropType(BackpropType.TruncatedBPTT)
                .tBPTTForwardLength(tbpttLength).tBPTTBackwardLength(tbpttLength)
                .pretrain(false)
                .build();
}
```

**Comment by AlexDBlack** [comment]
Can you post the shape of the input and labels arrays (literally just the output of INDArray.shape() for those arrays) that you are feeding into the net, to produce that issue? I might need to run this locally to work out what's going on there.

**Comment by Mioh** [comment]
Input shape: `[100, 1, 6, 20]`
Label shape: `[100, 9947]`

**Comment by Mioh** [comment]
Two notes.
1. Adding a dense layer in-between the CNN and RNN layers allows the network to run truncated BPTT.
2. Using tBPTTForwardLength = 1 and tBPTTBackwardLength =1 allows the network to train, but its behavior is strange. Epochs are completed without any iterations printed by the ScoreIterationListener and nothing shows up in the UI either.

**Comment by AlexDBlack** [comment]
I think this is a misconfiguration issue. Can you describe the architecture you are wanting to use in more detail, and specifically how the CNN input data relates to the RNN component? What is the "time" aspect here?
I had a quick look at the paper you linked, but couldn't work out what you were doing from that.

Another thing: with an RNN output layer, you want 3d (time series) data, whereas your labels are 2d.

**Comment by Mioh** [comment]
That's great, thank you for looking it to this. Hopefully you're right in that it is just a configuration issue. Actually, I made some faulty assumptions reading the article, however after looking at the authors [lua source code](https://github.com/yoonkim/lstm-char-cnn) and reading your feedback on RNN output I made some corrections. 

So, I'm trying to construct a Language Model which takes a time series of words represented as images with dimensions `[maxWordLength, characterEmbeddingLength] `(and depth=1), where the character embedding is just a one-hot representation of all the valid characters in my encoding.  The networks output is the probability distribution for the next word following the input sequence.
So the approach is somewhat similar to the [Shakespeare example](https://github.com/deeplearning4j/dl4j-examples/tree/master/dl4j-examples/src/main/java/org/deeplearning4j/examples/recurrent/character) except that predictions are made on a word level and each input is processed by the CNN before it is passed to the LSTM.

**Simplest architecture:** `InputLayer`->`ConvolutionLayer`->`SubsamplingLayer`->`GravesLSTM`->`RnnOutputLayer`
**Input:** `[minibatchSize * timeSeriesLength, 1, maxWordLength, characterEmbeddingLength]`
**Output:** `[miniBatchSize, outputVocabularySize, timeSeriesLength]`

**Note:** Since CNNs do not handle time series data, I'm using `minibatchSize * timeSeriesLength` as `minibachSize` for the input tensor to get the desired effect, as suggested by the [pre processor tests](https://github.com/deeplearning4j/deeplearning4j/blob/311c15fb2f6720a3e3dd3f3dba4ce468aa330202/deeplearning4j-core/src/test/java/org/deeplearning4j/nn/conf/preprocessor/TestPreProcessors.java).

I still get an error however.
`java.lang.IllegalStateException: Unable to get number of of rows for a non 2d matrix
	at NeuralLanguageModelTrainer.Trainer.train(Trainer.java:102)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.rows(BaseNDArray.java:3815)
	at NeuralLanguageModelTrainer.Trainer.main(Trainer.java:71)
	at org.nd4j.linalg.api.ndarray.BaseNDArray.mmul(BaseNDArray.java:2639)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at org.deeplearning4j.nn.layers.recurrent.LSTMHelpers.activateHelper(LSTMHelpers.java:170)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.activateHelper(GravesLSTM.java:150)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.deeplearning4j.nn.layers.recurrent.GravesLSTM.rnnActivateUsingStoredState(GravesLSTM.java:222)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.deeplearning4j.nn.graph.ComputationGraph.rnnActivateUsingStoredState(ComputationGraph.java:2081)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)
	at org.deeplearning4j.nn.graph.ComputationGraph.computeGradientAndScore(ComputationGraph.java:955)
	at org.deeplearning4j.optimize.solvers.BaseOptimizer.gradientAndScore(BaseOptimizer.java:151)
	at org.deeplearning4j.optimize.solvers.StochasticGradientDescent.optimize(StochasticGradientDescent.java:54)
	at org.deeplearning4j.optimize.Solver.optimize(Solver.java:51)
	at org.deeplearning4j.nn.graph.ComputationGraph.doTruncatedBPTT(ComputationGraph.java:2032)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:824)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:798)
	at org.deeplearning4j.nn.graph.ComputationGraph.fit(ComputationGraph.java:660)
	at org.deeplearning4j.earlystopping.trainer.EarlyStoppingGraphTrainer.fit(EarlyStoppingGraphTrainer.java:74)
	at org.deeplearning4j.earlystopping.trainer.BaseEarlyStoppingTrainer.fit(BaseEarlyStoppingTrainer.java:109)
	at NeuralLanguageModelTrainer.Trainer.train(Trainer.java:98)
	at NeuralLanguageModelTrainer.Trainer.main(Trainer.java:71)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:147)`

**Comment by Mioh** [comment]
This issue seems to be solved by the introduction of GlobalPoolingLayer in version 0.8.0 as what I'm really trying to achieve here is max pooling over time.

**Comment by Ritaplfigueiredo** [comment]
Hello @Mioh could you describe in more detail the achitecture that solved your issue? Getting similar error trying to combine CNN with LSTM. Thank you in advance,

Rita
